src/rust/generator-dispatcher/src/generator_ids_cache.rs

```diff
@@ -0,0 +1,298 @@
+    PluginRegistryClientError(#[from] PluginRegistryServiceClientError),
+}
+
+#[derive(Debug, Error)]
+pub enum GeneratorIdsCacheError {
+    #[error("fatal error {0}")]
+    Fatal(String),
+
+    #[error("retryable error {0}")]
+    Retryable(String),
+}
+
+impl From<TrySendError<Uuid>> for GeneratorIdsCacheError {
+    fn from(try_send_error: TrySendError<Uuid>) -> Self {
+        if try_send_error.is_full() {
+            GeneratorIdsCacheError::Retryable("update queue is full".to_string())
+        } else if try_send_error.is_disconnected() {
+            GeneratorIdsCacheError::Fatal("update queue is disconnected".to_string())
+        } else {
+            GeneratorIdsCacheError::Retryable(format!("unknown TrySendError: {}", try_send_error,))
+        }
+    }
+}
+
+struct GeneratorIdsEntry {
+    generator_ids: Vec<Uuid>,
+    evict_after: SystemTime,
+}
+
+impl GeneratorIdsEntry {
+    fn new(generator_ids: Vec<Uuid>, ttl: Duration) -> Self {
+        GeneratorIdsEntry {
+            generator_ids,
+            evict_after: SystemTime::now() + ttl,
+        }
+    }
+}
+
+/// A fail-fast, asynchronous, pull-through cache.
+///
+/// Caches a mapping of {<event_source_id>: [<generator_id>, ...]} in such a way
+/// that lookups return immediately, even on cache miss. This may seem
+/// counterproductive, but the idea is to use this in a Kafka consumer where we
+/// want to avoid waiting for an RPC round-trip while we're holding onto a
+/// message. Instead, we dump the message on a retry topic and asynchronously
+/// update the cache. That's why this cache returns immediately on both a hit
+/// and a miss.
+#[derive(Clone, Debug)]
+pub struct GeneratorIdsCache {
+    generator_ids_cache: Arc<RwLock<LruCache<Uuid, GeneratorIdsEntry>>>,
+    updater_tx: Sender<Uuid>,
+}
+
+impl GeneratorIdsCache {
+    /// Construct a generator IDs cache.
+    ///
+    /// # Arguments
+    ///
+    /// * capacity - the maximum number of keys the cache can hold at any time.
+    ///
+    /// * ttl - how long to wait before considering a cache entry "stale".
+    ///
+    /// * updater_pool_size - the maximum number of concurrent tasks for
+    ///   handling cache updates.
+    ///
+    /// * updater_queue_depth - the maximum number of waiting cache updates.
+    pub async fn new(
+        capacity: usize,
+        ttl: Duration,
+        updater_pool_size: usize,
+        updater_queue_depth: usize,
+    ) -> Result<Self, ConfigurationError> {
+        let generator_ids_cache: Arc<RwLock<LruCache<Uuid, GeneratorIdsEntry>>> =
+            Arc::new(RwLock::new(LruCache::new(capacity)));
+        let evictor_generator_ids_cache = generator_ids_cache.clone();
+        let updater_generator_ids_cache = generator_ids_cache.clone();
+
+        let (updater_tx, updater_rx) = futures::channel::mpsc::channel(updater_queue_depth);
+
+        // The evictor task is responsible for periodically evicting old entries
+        // from the cache.
```

Comment:
I believe moka has support for ttl natively

Reply:
fixed